### PR TITLE
added drive package

### PR DESCRIPTION
**Pull request by bobvanderlinden**
I used the same package structure as [packer](https://github.com/NixOS/nixpkgs/blob/0761f81da71fc6a940c7f51129b6c7717db78e87/pkgs/development/tools/packer/default.nix). This adds a command-line tool for Google Drive called 'drive'.

### CODE REVIEW

**Comment by copumpkin** [comment]
There's a whole load of go build infrastructure in `goPackages`. I wouldn't base your new expression off of packer, which I suspect was written before `goPackages` or perhaps @cstrahan wasn't aware of it when he did it. Seems like it would simplify your stuff significantly to just use `buildGoPackage` and the prepackaged dependencies (which you may need to add to, but then other people can use them too)

**Comment by bobvanderlinden** [comment]
Ah, I started off doing that, but it took quite some time by hand. After finding `packer` I thought go2nix was the way to go (since it saved a lot of time ;)). I'll have a look at using `go-packages.nix` again.

**Comment by copumpkin** [comment]
Yeah, we should probably just figure out if we can get a smooth auto-packaging mechanism for the `goPackages` system. Having said that, lots of your dependencies appear to already be in `goPackages` so this particular case doesn't seem that bad. Did anything in particular trip you up?

**Comment by bobvanderlinden** [comment]
Yes, I'm stuck on `grpc`, which is failing for me and also on Hydra: https://hydra.nixos.org/build/22490000/nixlog/1/tail-reload

It seems grpc needs to depend on gcloud-golang, but gcloud-golang already depends on grpc.
I tried to include the source of gcloud-golang, like oauth2 does:
```
    extraSrcs = [
      { inherit (gcloud-golang) src goPackagePath; }
    ];
```

But that causes other problems.

Is there a way to depend on only a subPackage of gcloud-golang? Or should this be solved another way?

**Comment by copumpkin** [comment]
I think @wkennington dealt with that problem recently, but don't know how he resolved it.

**Comment by bobvanderlinden** [comment]
https://github.com/NixOS/nixpkgs/pull/8362 should solve the issue. Not sure if it is the right way to go, but makes grpc buildable.

**Comment by bobvanderlinden** [comment]
I've made `drive` buildable with `buildGoPackage`.
e5d9bd2 is the commit from #8362, which fixes `grpc`.
216f29b adds the package with `buildGoPackage` and adds the dependencies to `goPackages`.
After these commits `drive` will not yet build. I got a strange permission denied problem related `oauth2`: permission denied upon installing through go. I fixed this issue by using `propagatedBuildInputs` instead of `extraSrcs`. I'm not 100% sure this is the right way to go, but it seemed cleaner. This is commit 67cad28. Please let me know if there is a better solution.

If this seems like good changes, I'll wait for #8362 to be merged and remove e5d9bd2 from this PR before this is ready .

File: pkgs/top-level/go-packages.nix [hunk 1]
```diff
@@ -1570,6 +1651,18 @@ let
     };
   };
 
+  log = buildGoPackage rec {
```
Comment (lethalman) [inline pkgs/top-level/go-packages.nix]: "log" is quite a common name, can you please use `odeke-em.log`.

**Comment by lethalman** [comment]
I've merged the grpc fix, can you rebase on current master?

**Comment by bobvanderlinden** [comment]
@lethalman thanks. I've rebased and renamed `log` to `odeke-em.log`.

**Comment by lethalman** [comment]
Built locally, thanks.

**Pull request merged**